apps/vlasov.c

```diff
@@ -384,31 +395,80 @@ vm_find_fluid_species_idx(const gkyl_vlasov_app *app, const char *nm)
   return -1;
 }
 
+void
+vm_apply_bc(gkyl_vlasov_app* app, double tcurr,
+  struct gkyl_array *distf[], struct gkyl_array *fluid[], struct gkyl_array *emfield)
+{
+  for (int i=0; i<app->num_species; ++i) {
+    vm_species_apply_bc(app, &app->species[i], distf[i], tcurr);
+  }
+  for (int i=0; i<app->num_fluid_species; ++i) {
+    vm_fluid_species_apply_bc(app, &app->fluid_species[i], fluid[i]);
+  }
+  if (app->has_field) {
+    if (!(app->field->field_id == GKYL_FIELD_PHI || app->field->field_id == GKYL_FIELD_PHI_EXT))
+      vm_field_apply_bc(app, app->field, emfield);
+  }
+}
+
+void
+vp_calc_field(gkyl_vlasov_app* app, double tcurr, const struct gkyl_array *fin[])
+{
+  // Compute electrostatic potential from Poisson's equation.
+  vp_field_accumulate_charge_dens(app, app->field, fin);
+
+  // Solve the field equation.
+  vp_field_solve(app, app->field);
+}
+
+void
+vp_calc_field_and_apply_bc(gkyl_vlasov_app* app, double tcurr, struct gkyl_array *distf[])
+{
+  // Compute the field.
+  vp_calc_field(app, tcurr, (const struct gkyl_array **) distf);
+
+  // Apply boundary conditions.
+  for (int i=0; i<app->num_species; ++i) {
+    vm_species_apply_bc(app, &app->species[i], distf[i], tcurr);
+  }
+}
+
 void
 gkyl_vlasov_app_apply_ic(gkyl_vlasov_app* app, double t0)
 {
   app->tcurr = t0;
-  if (app->has_field) 
-    gkyl_vlasov_app_apply_ic_field(app, t0);
   for (int i=0; i<app->num_species; ++i)
     gkyl_vlasov_app_apply_ic_species(app, i, t0);
-  // BCs must be done after all species initialize for emission BCs to work 
+
+  if (app->has_field) 
+    gkyl_vlasov_app_apply_ic_field(app, t0);
+
+  struct gkyl_array *distf[app->num_species];
+  struct gkyl_array *fluid[app->num_fluid_species];
   for (int i=0; i<app->num_species; ++i)
-    vm_species_apply_bc(app, &app->species[i], app->species[i].f, t0);
+    distf[i] = app->species[i].f;
   for (int i=0; i<app->num_fluid_species; ++i)
-    gkyl_vlasov_app_apply_ic_fluid_species(app, i, t0);
+    fluid[i] = app->fluid_species[i].fluid;
+  // BCs must be done after all species initialize for emission BCs to work.
+  vm_apply_bc(app, t0, distf, fluid, app->has_field ? app->field->em : 0);
 }
 
 void
 gkyl_vlasov_app_apply_ic_field(gkyl_vlasov_app* app, double t0)
 {
   app->tcurr = t0;
-
   struct timespec wtm = gkyl_wall_clock();
-  vm_field_apply_ic(app, app->field, t0);
-  app->stat.init_field_tm += gkyl_time_diff_now_sec(wtm);
 
-  vm_field_apply_bc(app, app->field, app->field->em);
+  if (app->field->field_id == GKYL_FIELD_PHI || app->field->field_id == GKYL_FIELD_PHI_EXT) {
+    struct gkyl_array *distf[app->num_species];
+    for (int i=0; i<app->num_species; ++i)
+      distf[i] = app->species[i].f;
+    vp_calc_field(app, t0, (const struct gkyl_array **) distf);
```

Comment:
What is the cast to a const struct gkyl_array ** for? Shouldn't the function signature be for a gkyl_array* f[] so you can just pass distf (like what we do for passing distribution functions to the forward Euler methods).

Reply:
if you don't cast it doesn't work. You can try it. You'll get a cascade of error messages and warnings, and as you try to resolve them you'll find that all the arrays of gkyl_arrays (e.g. const struct gkyl_array *fin[]) need to be defined without a const throughout the whole system (e.g. in vlasov_forward_euler), in both VM and VP. This casting is a minimal modification approach, and is also what we used in GK.